### DISCUSSION TITLE
Add Filtering to work/digest Option-values

### DISCUSSION THREAD

**Discussion by MaxLuchs** [diesel-rs/diesel#2656] [Ideas]
Hello, it would be super cool, if Diesel's-Filter-method on Table would work with Rust-Option as well. That would make dynamic Filtering much easier.

`schema::dogs::filter(name.eq(Some("snoopy")).and(species.eq(None)))`

And if None is provided, the specific condition is skipped: `species.eq(None))`

**Comment by MaxLuchs** [comment]
Or maybe that is not possible because Rust.-None is always interpreted as SQL-null

**Reply by weiznich** [reply]
That's correct + a bunch of other performance related reasons. Basically the same reasoning from https://github.com/diesel-rs/diesel/issues/1306 also applies here.

That written: Such a implementation must not necessarily be done in diesel itself, any third party crate could extend our query ast and provide such a method.

**Reply by MaxLuchs** [reply]
Thanks. As an alternative Diesel could use Structs for where-Filtering as well. Similar to Changeset or Insertable. Then you could annotate fields, if None should map to null or discarded.
In MongoDB it works that way and you pass JS-Objects as Queries. But that might alienate Diesel from SQL maybe on the other hand. But would find it handy even so.
```
#[derive(Filterable)]
struct UserFilter {
     name: String,
     #[discardIfNone]
     age: Option<i32>,
}

user::filter(user_filter).get_results::<Vec<User>>()
```
In this way you could re-use the same generic Diesel-Query-Statement and configure the Query-statement by passing different data as Filter into it w/o having to change the Query.

**Reply by weiznich** [reply]
There is no need for having such an abstraction in diesel itself, as something like this could easily be implemented as third party crate on top of diesel. You would need to provide an `AsExpression` impl for such a type + build an `Box<dyn BoxableExpression>` internally.

**Reply by MaxLuchs** [reply]
Cool that Diesel is so extendable. Need to learn more Rust then since I dont know how to write macros yet. Though I feel like using Filter-structs would be more consistent with how Diesel handled already other parts of interacting with SQL via Structs. Where you have also the choice as a user between Tuples + Structs. Just my thoughts and tiny nitpicking. Great framework for sure
